File: crates/lanekeep-rules/rules/no-circular-imports.ts
import { defineRule } from 'lanekeep'
import { resolveImport } from 'lanekeep/paths'

/**
 * Import cycles.
 *
 * A cycle is the architectural failure that hides best. Everything compiles, the bundler
 * copes, and then one module observes a half-initialized binding from another and produces
 * an error miles from its cause — usually only under a particular import order, which is
 * why it survives review and shows up in production.
 *
 * No single file contains a cycle, so this is a reduce-phase rule by nature.
 *
 * @example
 * ```ts
 * import noCircularImports from 'lanekeep/no-circular-imports'
 *
 * export default defineConfig({
 *   rules: [noCircularImports({ maxDepth: 12 })],
 * })
 * ```
 */
export default function noCircularImports(options) {
  // A cycle spanning more files than this is not reported. Cycles get harder to act on the
  // longer they are, and an unbounded search on a pathological graph is the one way this
  // rule could become the slowest thing in a run.
  const maxDepth = options?.maxDepth ?? 24

  return defineRule({
    id: 'lanekeep/no-circular-imports',
    severity: 'error',

    card: {
      message: 'circular import',
      remediation:
        'break the cycle — extract what both modules need into a third, or invert one dependency',
      examples: {
        bad: "// a.ts imports b.ts, and b.ts imports a.ts",
        good: "// a.ts and b.ts both import shared.ts",
      },
    },

    timeout: 5_000,

    query: `
      (import_statement source: (string) @source) @stmt
      (export_statement source: (string) @source) @stmt
    `,

    check(ctx, m) {
      ctx.emitFact({
        kind: 'edge',
        to: ctx.text(m.source).slice(1, -1),
        line: ctx.line(m.stmt),
        column: ctx.column(m.stmt),
      })
    },

    reduce(ctx) {
      const files = new Set(ctx.files)

      // file -> [{ to, line, column }], in fact order, which is (file, emission) order.
      // Everything downstream inherits that determinism.
      const edges = new Map()
      for (const e of ctx.facts('edge')) {
        const target = resolveImport(e.file, e.to, files)
        if (!target || target === e.file) continue

        if (!edges.has(e.file)) edges.set(e.file, [])
        edges.get(e.file).push({ to: target, line: e.line, column: e.column })
      }

      const cycles = findCycles(edges, maxDepth)

      for (const cycle of cycles) {
        // Reported at the import that closes the cycle — the one edge whose removal breaks
        // it. Reporting at every member would turn one problem into N violations and leave
        // the reader to work out they are the same thing.
        ctx.report(
          { file: cycle.file, line: cycle.line, column: cycle.column },
          `circular import: ${cycle.path.join(' → ')}`,
        )
      }
    },
  })
}

/**
 * Every distinct cycle in the graph, each reported once.
 *
 * Iterative depth-first search rather than recursion: a corpus deep enough to overflow the
 * stack would abort the whole run, and a rule must not be able to do that to a run by being
 * given a large repository.
 */
function findCycles(edges, maxDepth) {
  const found = []
  const seen = new Set()

  // Sorted, so the search visits roots in the same order regardless of how the fact list
  // happened to group. Fact order already gives that, but this does not depend on it.
  const roots = [...edges.keys()].sort()

  for (const root of roots) {
    // Each frame: the file, and how far into its edge list we have walked.
    const stack = [{ file: root, index: 0 }]
    const onPath = new Map([[root, 0]])

    while (stack.length > 0) {
      const frame = stack[stack.length - 1]
      const outgoing = edges.get(frame.file) ?? []

      if (frame.index >= outgoing.length) {
        stack.pop()
        onPath.delete(frame.file)
        continue
      }

      const edge = outgoing[frame.index]
      frame.index += 1

      const at = onPath.get(edge.to)
      if (at !== undefined) {
        // This edge closes a cycle: everything on the path from `edge.to` onward.
        const path = stack.slice(at).map((f) => f.file)
        const key = canonical(path)
        if (!seen.has(key)) {
          seen.add(key)
          found.push({
            file: frame.file,
            line: edge.line,
            column: edge.column,
            path: [...path, edge.to],
          })
        }
        continue
      }

      if (stack.length >= maxDepth) continue

      stack.push({ file: edge.to, index: 0 })
      onPath.set(edge.to, stack.length - 1)
    }
  }

  // Sorted by where the violation lands, so two runs over the same corpus report the same
  // cycles in the same order even if the search ever changes.
  found.sort(
    (a, b) => compare(a.file, b.file) || a.line - b.line || a.column - b.column,
  )
  return found
}

/**
 * A cycle's identity, independent of which member the search entered from.
 *
 * `a → b → a` and `b → a → b` are one cycle. Without this the same problem is reported once
 * per member, which is both noise and a different count on a graph that has not changed.
 */
function canonical(path) {
  let best = null
  for (let i = 0; i < path.length; i += 1) {
    const rotation = [...path.slice(i), ...path.slice(0, i)].join('\u0000')
    if (best === null || rotation < best) best = rotation
  }
  return best ?? ''
}

/** String ordering that does not depend on a locale the sandbox does not have. */
function compare(a, b) {
  if (a < b) return -1
  return a > b ? 1 : 0
}

File: crates/lanekeep-cli/tests/no_circular_imports.rs
//! `lanekeep/no-circular-imports`, run through the binary over a real corpus.

#![expect(
    clippy::expect_used,
    clippy::panic,
    reason = "`clippy.toml`'s allow-*-in-tests only reaches `#[test]` functions and \
              `#[cfg(test)]` modules. The `corpus` helpers are neither, so the grant it \
              already makes for unit tests has to be restated for them."
)]

mod corpus;

use corpus::Corpus;

fn corpus(options: &str, files: &[(&str, &str)]) -> Corpus {
    Corpus::new("no-circular-imports", options, files)
}

#[test]
fn an_acyclic_graph_reports_nothing() {
    let found = corpus(
        "{}",
        &[
            (
                "src/a.ts",
                "import { b } from './b';\nexport const a = b;\n",
            ),
            (
                "src/b.ts",
                "import { c } from './c';\nexport const b = c;\n",
            ),
            ("src/c.ts", "export const c = 1;\n"),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_two_file_cycle_is_reported() {
    let found = corpus(
        "{}",
        &[
            (
                "src/a.ts",
                "import { b } from './b';\nexport const a = b;\n",
            ),
            (
                "src/b.ts",
                "import { a } from './a';\nexport const b = a;\n",
            ),
        ],
    )
    .run();
    assert_eq!(found.len(), 1, "one cycle, one violation: {found:?}");
    assert!(
        found[0].contains("circular import: src/a.ts → src/b.ts → src/a.ts"),
        "{found:?}"
    );
}

#[test]
fn a_longer_cycle_is_reported_once() {
    // Three members, one problem. Reporting per member would turn one cycle into three
    // violations and leave the reader to work out they are the same thing.
    let found = corpus(
        "{}",
        &[
            (
                "src/a.ts",
                "import { b } from './b';\nexport const a = b;\n",
            ),
            (
                "src/b.ts",
                "import { c } from './c';\nexport const b = c;\n",
            ),
            (
                "src/c.ts",
                "import { a } from './a';\nexport const c = a;\n",
            ),
        ],
    )
    .run();
    assert_eq!(found.len(), 1, "{found:?}");
    assert!(
        found[0].contains("src/a.ts → src/b.ts → src/c.ts → src/a.ts"),
        "{found:?}"
    );
}

#[test]
fn two_independent_cycles_are_both_reported() {
    let found = corpus(
        "{}",
        &[
            (
                "src/a.ts",
                "import { b } from './b';\nexport const a = b;\n",
            ),
            (
                "src/b.ts",
                "import { a } from './a';\nexport const b = a;\n",
            ),
            (
                "src/x.ts",
                "import { y } from './y';\nexport const x = y;\n",
            ),
            (
                "src/y.ts",
                "import { x } from './x';\nexport const y = x;\n",
            ),
        ],
    )
    .run();
    assert_eq!(found.len(), 2, "{found:?}");
}

#[test]
fn a_self_import_is_not_a_cycle() {
    // A module importing itself is a different mistake, and reporting it here would be
    // confusing advice — there is no second module to extract anything into.
    let found = corpus(
        "{}",
        &[(
            "src/a.ts",
            "import { a } from './a';\nexport const b = a;\n",
        )],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_diamond_is_not_a_cycle() {
    // Two paths to the same module is ordinary structure. A search that confused a
    // revisited node with a node on the current path would report this.
    let found = corpus(
        "{}",
        &[
            (
                "src/top.ts",
                "import { l } from './left';\nimport { r } from './right';\nexport const t = l + r;\n",
            ),
            ("src/left.ts", "import { s } from './shared';\nexport const l = s;\n"),
            ("src/right.ts", "import { s } from './shared';\nexport const r = s;\n"),
            ("src/shared.ts", "export const s = 1;\n"),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_re_export_edge_counts() {
    // `export ... from` is an import edge with different syntax, and a cycle through one
    // fails at runtime exactly the same way.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", "export { b } from './b';\n"),
            (
                "src/b.ts",
                "import { a } from './a';\nexport const b = a;\n",
            ),
        ],
    )
    .run();
    assert_eq!(found.len(), 1, "{found:?}");
}

#[test]
fn a_package_import_is_not_an_edge() {
    let found = corpus(
        "{}",
        &[
            (
                "src/a.ts",
                "import merge from 'lodash';\nexport const a = merge;\n",
            ),
            (
                "src/b.ts",
                "import path from 'node:path';\nexport const b = path;\n",
            ),
        ],
    )
    .run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn max_depth_bounds_the_search() {
    // A cycle longer than the bound is not reported. Deliberate: an unbounded search on a
    // pathological graph is the one way this rule could dominate a run.
    let files: Vec<(String, String)> = (0..6)
        .map(|i| {
            let next = (i + 1) % 6;
            (
                format!("src/m{i}.ts"),
                format!("import {{ m{next} }} from './m{next}';\nexport const m{i} = m{next};\n"),
            )
        })
        .collect();
    let layout: Vec<(&str, &str)> = files
        .iter()
        .map(|(p, c)| (p.as_str(), c.as_str()))
        .collect();

    let generous = Corpus::new("no-circular-imports", "{ maxDepth: 24 }", &layout).run();
    assert_eq!(
        generous.len(),
        1,
        "the six-file cycle should be found: {generous:?}"
    );

    let tight = Corpus::new("no-circular-imports", "{ maxDepth: 3 }", &layout).run();
    assert_eq!(
        tight,
        Vec::<String>::new(),
        "the bound should stop the search"
    );
}

#[test]
fn the_same_corpus_reports_the_same_thing_every_run() {
    // Determinism at the level a reader sees: same cycles, same order, same anchors.
    let corpus = corpus(
        "{}",
        &[
            (
                "src/a.ts",
                "import { b } from './b';\nexport const a = b;\n",
            ),
            (
                "src/b.ts",
                "import { a } from './a';\nexport const b = a;\n",
            ),
            (
                "src/x.ts",
                "import { y } from './y';\nexport const x = y;\n",
            ),
            (
                "src/y.ts",
                "import { z } from './z';\nexport const y = z;\n",
            ),
            (
                "src/z.ts",
                "import { x } from './x';\nexport const z = x;\n",
            ),
        ],
    );
    let first = corpus.run();
    assert_eq!(first.len(), 2, "{first:?}");
    for attempt in 0..4 {
        assert_eq!(corpus.run(), first, "output changed on attempt {attempt}");
    }
}
